pandas/tests/series/test_analytics.py

```diff
@@ -2028,6 +2030,33 @@ def test_n(self, n):
         expected = s.sort_values().head(n)
         assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize('dtype', [
+        'int8', 'int16', 'int32', 'int64',
+        'uint8', 'uint16', 'uint32', 'uint64',
+        'float16', 'float32', 'float64',
+        'datetime64[ns]', 'timedelta64[ns]'])
+    @pytest.mark.parametrize('method', ['nsmallest', 'nlargest'])
+    def test_boundary(self, method, dtype):
+        # GH 21426
+        if is_float_dtype(dtype):
+            min_val, max_val = np.finfo(dtype).min, np.finfo(dtype).max
+            min_2nd, max_2nd = np.nextafter([min_val, max_val], 0, dtype=dtype)
+            vals = [min_val, min_2nd, max_2nd, max_val]
+        elif is_integer_dtype(dtype):
+            min_val, max_val = np.iinfo(dtype).min, np.iinfo(dtype).max
+            vals = [min_val, min_val + 1, max_val - 1, max_val]
+        elif is_datetimelike(dtype):
+            # use int64 bounds and +1 to min_val since true minimum is NaT
+            # (include min_val/NaT at end to maintain same expected_idxr)
+            min_val, max_val = np.iinfo('int64').min, np.iinfo('int64').max
+            vals = [min_val + 1, min_val + 2, max_val - 1, max_val, min_val]
+
+        s = Series(vals, dtype=dtype)
+        result = getattr(s, method)(3)
+        expected_idxr = [0, 1, 2] if method == 'nsmallest' else [3, 2, 1]
+        expected = s.loc[expected_idxr]
+        tm.assert_series_equal(result, expected)
```

Comment:
I like the thought of using parameterization, but this is a little hard to read.  I would recommend the following:

* Split this test into three tests based on which conditional it hits.
* Extract lines 2054-2058 as a separate (non-test) method, where you pass in `vals` as a parameter i.e.

~~~python
def _check_nselect(self, vals, method):
   s = Series(...)
   ...
   tm.assert_series_equal(result, expected)
~~~

Reply:
done